### PR TITLE
Specialize interleave string ~2-3x faster

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #2864

# Rationale for this change
 

```
interleave str(20, 0.0) 100 [0..100, 100..230, 450..1000]
                        time:   [1.0792 µs 1.0795 µs 1.0799 µs]
                        change: [-60.193% -60.122% -60.019%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe

interleave str(20, 0.0) 400 [0..100, 100..230, 450..1000]
                        time:   [3.2179 µs 3.2187 µs 3.2197 µs]
                        change: [-60.739% -60.622% -60.497%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) low mild
  2 (2.00%) high mild
  5 (5.00%) high severe

interleave str(20, 0.0) 1024 [0..100, 100..230, 450..1000]
                        time:   [7.4057 µs 7.4079 µs 7.4101 µs]
                        change: [-60.606% -60.542% -60.442%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

interleave str(20, 0.0) 1024 [0..100, 100..230, 450..1000, 0..1000]
                        time:   [7.4336 µs 7.4369 µs 7.4403 µs]
                        change: [-58.701% -58.630% -58.525%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  7 (7.00%) high mild
  2 (2.00%) high severe

interleave str(20, 0.5) 100 [0..100, 100..230, 450..1000]
                        time:   [1.6810 µs 1.6815 µs 1.6821 µs]
                        change: [-58.572% -58.434% -58.308%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  2 (2.00%) high severe

interleave str(20, 0.5) 400 [0..100, 100..230, 450..1000]
                        time:   [5.2158 µs 5.2246 µs 5.2341 µs]
                        change: [-64.090% -63.972% -63.864%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

interleave str(20, 0.5) 1024 [0..100, 100..230, 450..1000]
                        time:   [16.993 µs 17.013 µs 17.032 µs]
                        change: [-54.202% -54.077% -53.991%] (p = 0.00 < 0.05)
                        Performance has improved.

interleave str(20, 0.5) 1024 [0..100, 100..230, 450..1000, 0..1000]
                        time:   [17.424 µs 17.443 µs 17.464 µs]
                        change: [-52.353% -52.256% -52.128%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
```

# What changes are included in this PR?

Adds a specialized implementation for string arrays

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by tustvold** [comment]
With #2947 this will be generalizable to also cover byte arrays

File: arrow-array/src/array/string_array.rs [hunk 1]
```diff
@@ -103,10 +103,8 @@ impl<OffsetSize: OffsetSizeTrait> GenericStringArray<OffsetSize> {
         // OffsetSizeTrait. Currently, only i32 and i64 implement OffsetSizeTrait,
         // both of which should cleanly cast to isize on an architecture that supports
         // 32/64-bit offsets
-        let slice = std::slice::from_raw_parts(
-            self.value_data.as_ptr().offset(start.to_isize().unwrap()),
-            (*end - *start).to_usize().unwrap(),
-        );
+        let slice =
+            std::slice::from_raw_parts(self.value_data.as_ptr().add(start), end - start);
```
Comment (alamb) [inline arrow-array/src/array/string_array.rs:107]: this is a drive by cleanup right?

File: arrow-select/src/interleave.rs [hunk 2]
```diff
@@ -85,51 +85,105 @@ pub fn interleave(
...
     downcast_primitive! {
         data_type => (primitive_helper, values, indices, data_type),
+        DataType::Utf8 => interleave_string::<i32>(values, indices, data_type),
+        DataType::LargeUtf8 => interleave_string::<i64>(values, indices, data_type),
         _ => interleave_fallback(values, indices)
     }
 }
 
+/// Common functionality for interleaving arrays
+struct Interleave<'a, T> {
```
Comment (alamb) [inline arrow-select/src/interleave.rs:95]: some comments might help here specifically what `null_count` and `nulls` represent and what the generic `T` is used for

File: arrow-select/src/interleave.rs [hunk 3]
```diff
@@ -85,51 +85,105 @@ pub fn interleave(
...
 ) -> Result<ArrayRef, ArrowError> {
-    let mut has_nulls = false;
-    let cast: Vec<_> = values
-        .iter()
-        .map(|x| {
-            has_nulls = has_nulls || x.null_count() != 0;
-            as_primitive_array::<T>(*x)
-        })
-        .collect();
+    let interleaved = Interleave::<'_, PrimitiveArray<T>>::new(values, indices);
```
Comment (alamb) [inline arrow-select/src/interleave.rs:136]: 👍

File: arrow-select/src/interleave.rs [hunk 4]
```diff
@@ -85,51 +85,105 @@ pub fn interleave(
...
+    data_type: &DataType,
+) -> Result<ArrayRef, ArrowError> {
+    let interleaved = Interleave::<'_, GenericStringArray<O>>::new(values, indices);
+
+    let mut capacity = 0;
+    let mut offsets = BufferBuilder::<O>::new(indices.len() + 1);
+    offsets.append(O::from_usize(0).unwrap());
+    for (a, b) in indices {
+        let o = interleaved.arrays[*a].value_offsets();
+        let len = o[*b + 1].as_usize() - o[*b].as_usize();
```
Comment (alamb) [inline arrow-select/src/interleave.rs:166]:
```suggestion
        // element length
        let len = o[*b + 1].as_usize() - o[*b].as_usize();
```

File: arrow-select/src/interleave.rs [hunk 5]
```diff
@@ -85,51 +85,105 @@ pub fn interleave(
...
+    values: &[&dyn Array],
+    indices: &[(usize, usize)],
+    data_type: &DataType,
+) -> Result<ArrayRef, ArrowError> {
+    let interleaved = Interleave::<'_, GenericStringArray<O>>::new(values, indices);
+
+    let mut capacity = 0;
+    let mut offsets = BufferBuilder::<O>::new(indices.len() + 1);
+    offsets.append(O::from_usize(0).unwrap());
+    for (a, b) in indices {
```
Comment (alamb) [inline arrow-select/src/interleave.rs:164]: this is clever -- do the offsets in one pass and the strings in another

**Review by alamb — APPROVED** [review]
LGTM -- thank you

**Pull request merged**